### DISCUSSION TITLE
Support for audio recording

### DISCUSSION THREAD

**Discussion by kankadev** [Motion-Project/motion#1366] [General]
Hello,

I use motion as surveillance system and I'm missing the possibility to record audio with an USB microphone or with an USB webcam with builtin mic, too.

I tried to code something to start audio recording manually when the event `on_motion_detected` is triggered, but it's difficult to stop the audio recording when motion is over and to compare video and audio into one file.

**Comment by af7567** [comment]
Has anyone done any work on audio recording built into motion, rather than using on_motion_detected and _ended scripts? 

I had a quick look at the code the other day thinking I might spot somewhere I can just hack it in quickly to try it out, but soon realised it's not that easy to pass the input data to the output data. :)

I am more interested in recording from an RTSP stream than from USB, and while it can be done with scripts it seems wrong to open another RTSP connection to the camera just to grab the audio since that data is already being sent to motion. It also takes a few seconds for ffmpeg to be able to start capturing audio from a new connection even on the substream, so it would be better if it can be done from the already open stream.

**Comment by Mr-DaveDev** [comment]
The only "work" is within the [MotionPlus](https://github.com/Motion-Project/motionplus) application but it is not working very well.  There is an extreme latency.  I would never be putting this into Motion so you'd have to wait for the next maintainer if you want it inside of Motion.  

Everyone should also be acutely aware of the legal ramifications.  In "most" countries it is fine to record video but illegal to record audio security footage of people.  (To my knowledge, "most" includes: USA, Canada, Europe, Australia, NZ and probably many more).  It would be real bummer to submit the recording to the authorities for something happening to your property and then you are the one ending up with a felony for wiretapping!

**Reply by tv21** [reply]
I don't think that is true in the USA if you recording outdoors on your own property, and besides that, if that were true how do things like Ring doorbells get away with recording audio and video?  For that matter, if that were true even shooting video with a cell phone would be legally risky.  I think maybe you are confusing the laws that apply to employers  recording within workplaces with the laws that would apply to homeowners recording outdoors on their own property.  In general anything that happens outdoors has no expectation of privacy except under certain limited circumstances.  In any case I have never heard of anyone getting charged with so much as a misdemeanor for recording audio outdoors on their own property (and when I say their own property, I mean property they own or are currently renting).

**Reply by Mr-Dave** [reply]
You are incorrect.  Just google it and look at the legal web sites.

**Reply by gene-git** [reply]
Take all below as pure opinion/conjecture etc. Discuss with an appropriate attorney as necessary.

That said, its' basically fine with some restrictions. And there are federal as well as state laws that are relevant.

Federal:

US federal law (Specifically 18 U.S.C. § [2511] ) requires one party being recorded to be present during or alternatively the consent of any one-party. This is commonly called  'one-party consent'. 

State:

 In addition each state has it's own laws - the majority are one-party consent. 

There are  things which may suffice for consent  in some circumstances, beeps etc. And consent may be provided by a notice that recording devices are present (we've likely all seen these signs posted somewhere).

If no people are present, like birds or whatever - there are no restrictions at all. I suppose if person making recording speaks while recording a cell audio or in your home that would probably satisfy the one party consent requirement.  Public spaces where there is no expectation of privacy may be okay as well (at least for some states) - but I'm assuming that is largely not under discussion here.

**Reply by gene-git** [reply]
Forgot to add that places such as front porch area can be considered as having no expectation of privacy and that might be how 'recording' doorbells could be construed.  That seems to be an exception in (many) state statutes. 

That said this "no expectation of privacy exception"  may not exist under Federal wire tap laws (15 USC § 2511). 
A person doing this may be unlikely to be prosecuted, perhaps, as long as the usage is appropriately  limited - but it still may violate the federal statute.

Clear notification does seem to satisfy both state and federal laws and documenting such notification could be helpful too :)

**Reply by pfzim** [reply]
Why do you rely only on US laws? Sound recording may not be prohibited in other countries. All commercial products can record sound from cameras. 

Finally, make audio recording optional. Add sound recording at least from RTSP stream, I think it will not be difficult for you.

**Reply by gene-git** [reply]
My comment was obviously about US laws only. I am not familiar with those in other countries and Dave is the dev not me. 
I'm just a happy user :)

That said, maybe you should try MotionPlus. As @Mr-Dave said early on in this thread, work on sound has been happening there.
 
I can say I switched over from Motion to MotionPlus and audio is working just fine for me with rtsp streams - or at least as fine as the camera mic provides :)

**Reply by af7567** [reply]
I don't know why I never looked at MotionPlus before, I should give that a go :)
In the UK it is legal to record audio and video in your driveway/garden for peronal use. The only real problems would be if you share those recordings with third parties without getting permission from the people who were recorded.

**Reply by tv21** [reply]
> You are incorrect. Just google it and look at the legal web sites.

You're kidding, right?  Just google what, exactly?  And as I and others have pointed out, commercial products record sound so either there are no such laws or they aren't being enforced.  You even see videos (with sound) from doorbell cams featured in online videos and TV shows.  Note I am not talking about situations where an attempt is being made to hide the camera, but some of us use such cameras to monitor a front or side door and they are quite obvious if you are looking for them.

Anyway, Google is not a lawyer, and you probably shouldn't be taking legal advice from some random Google link.

**Reply by Mr-Dave** [reply]
Please go elsewhere.

**Comment by gene-git** [comment]
I also note that connecting to rtsp stream using VLC i get video and audio - however going through motion there is video only.
Obviously with multiple cameras on the main web page audio should be off for all but one (or none) - but when watching an individual camera it would be nice if audio was on - this is all audio + video over network no usb obviously.  

Is there something I'm have not set up correctly I wonder?

**Comment by gene-git** [comment]
Following on with audio vs video re cording - in the US video is generally fine, but its considered polite to inform. Audio rules/laws vary by state - some states are 2 party consent while some are 1 party - but even in the latter case, there must be 1 party - and if "you" are not present then its not okay. Once again, the standard is to inform that recording is happening.  There are things applicable to phone calls (periodic beeps for example) that do not make sense for surveillance - so one need ensure that all parties are made aware - then it's fine.

**Comment by meirlazar** [comment]
I've created a hacky solution for recording audio as a separate process, and it syncs up very well with the video
The motionaudio script will be called in the motion.conf file with 2 arguments (either start or stop the audio recording) and the 2nd argument uses the same format used for the movie_filename. This will generate a file.wav and file.pid while the file.avi (video file) is created by motion. The stop argument kills the arecord process and combines the audio and video file into a new file and removes the other 2 files.

## motion.conf
```
ffmpeg_video_codec mpeg4 # produces .avi files
movie_filename %m-%d-%Y_%H:%M:%S
on_event_start /bin/bash /path/to/motionaudio.sh start %m-%d-%Y_%H:%M:%S
on_event_end  /bin/bash /path/to/motionaudio.sh stop %m-%d-%Y_%H:%M:%S
```
# motionaudio.sh 
```
#!/bin/bash
moviedir=/path/to/motion/videos # change this 
action=$1
audio=$2
aext=wav # change if you prefer a different audio format or extension
vext=avi # change this to what is specified in the motion.conf file for the line ffmpeg_video_codec
voutext=mkv #change this to the video file that will be created by joining the audio and video files

function startaudio () {
  if pgrep -f "arecord"; then sudo pkill -9 -f "arecord -f cd -r 22050 -D plughw:2,0"; fi
  arecord -f cd -r 22050 -D plughw:2,0 "${moviedir}/${audio}.${aext}" &
  audiopid=${BASHPID}
  echo "${audiopid}" > "${moviedir}/${audio}.pid"
}

function stopaudio () {
  audiopids=$(find ${moviedir} -type f -iname "*.pid" -printf "%f\n" | cut -d. -f1)
  while read -r x; do 
	  if pgrep -F "${x}.pid"; then sudo pkill -9 -F "${x}.pid"; fi
	  if [[ -f "${moviedir}/${x}.${vext}" ]] && [[ -f "${moviedir}/${x}.${aext}" ]]; then
	  ffmpeg -y -i "${moviedir}/${x}.${vext}" -i "${moviedir}/${x}.${aext}" -c:v copy -c:a copy "${moviedir}/${x}.${voutext}" &&  rm -f "${moviedir}/${x}".{$aext,$vext,pid}
	  fi
  done <<< "${audiopids}"
  if ! pgrep -f "sudo nohup motion"; then sudo pkill -9 -f "arecord -f cd -r 22050 -D plughw:2,0"; fi
}

case ${action,,} in
stop ) stopaudio ;;
start ) startaudio ;;
esac
```

**Reply by ceewanna** [reply]
Based on your approach, I have tried recording parallel video files (with audio) with some loggings and found that the parallel videos are actually longer than those generated by motion. It is driven by event_gap. The on_event_end will kick in only when the event_gap has past which is way after the actual event ends. For instance, the default event_gap is 60s, then all the videos will have at least that event_gap built-in at the end. I am not so sure you have any problem reconcile your audio recorded with the intrinsic video generated by motion.

**Reply by ReedRYoung** [reply]
Nice work meirlazar, but when I tried to use motion to open a .WAV file in aplay "on_motion_detected", one of several problems I encountered was that PulseAudio didn't love references to specific sound devices, such as the "plughw:2,0" in your motionaudio.sh script.  So I'm wondering, is there a more hands-off way to pipe both the audio and video more directly to a software package that's intended for A/V recording, and let that software handle all or most of the details in your "startaudio" and "stopaudio" functions, except for filepath?

**Reply by meirlazar** [reply]
I've added this hack to a github project - feel free to download, test, and submit any feedback/updates. Thanks everyone.
I've made some changes to it, so hopefully that resolves most of the issues or complaints I've heard so far...
https://github.com/meirlazar/motion_with_audio_arecord

**Reply by drunkly** [reply]
> I've created a hacky solution for recording audio as a separate process, and it syncs up very well with the video The motionaudio script will be called in the motion.conf file with 2 arguments (either start or stop the audio recording) and the 2nd argument uses the same format used for the movie_filename. This will generate a file.wav and file.pid while the file.avi (video file) is created by motion. The stop argument kills the arecord process and combines the audio and video file into a new file and removes the other 2 files.
> 
> ## motion.conf
> ```
> ffmpeg_video_codec mpeg4 # produces .avi files
> movie_filename %m-%d-%Y_%H:%M:%S
> on_event_start /bin/bash /path/to/motionaudio.sh start %m-%d-%Y_%H:%M:%S
> on_event_end  /bin/bash /path/to/motionaudio.sh stop %m-%d-%Y_%H:%M:%S
> ```
> 
> # motionaudio.sh
> ```
> #!/bin/bash
> moviedir=/path/to/motion/videos # change this 
> action=$1
> audio=$2
> aext=wav # change if you prefer a different audio format or extension
> vext=avi # change this to what is specified in the motion.conf file for the line ffmpeg_video_codec
> voutext=mkv #change this to the video file that will be created by joining the audio and video files
> 
> function startaudio () {
>   if pgrep -f "arecord"; then sudo pkill -9 -f "arecord -f cd -r 22050 -D plughw:2,0"; fi
>   arecord -f cd -r 22050 -D plughw:2,0 "${moviedir}/${audio}.${aext}" &
>   audiopid=${BASHPID}
>   echo "${audiopid}" > "${moviedir}/${audio}.pid"
> }
> 
> function stopaudio () {
>   audiopids=$(find ${moviedir} -type f -iname "*.pid" -printf "%f\n" | cut -d. -f1)
>   while read -r x; do 
> 	  if pgrep -F "${x}.pid"; then sudo pkill -9 -F "${x}.pid"; fi
> 	  if [[ -f "${moviedir}/${x}.${vext}" ]] && [[ -f "${moviedir}/${x}.${aext}" ]]; then
> 	  ffmpeg -y -i "${moviedir}/${x}.${vext}" -i "${moviedir}/${x}.${aext}" -c:v copy -c:a copy "${moviedir}/${x}.${voutext}" &&  rm -f "${moviedir}/${x}".{$aext,$vext,pid}
> 	  fi
>   done <<< "${audiopids}"
>   if ! pgrep -f "sudo nohup motion"; then sudo pkill -9 -f "arecord -f cd -r 22050 -D plughw:2,0"; fi
> }
> 
> case ${action,,} in
> stop ) stopaudio ;;
> start ) startaudio ;;
> esac
> ```

I had some issues on low performance device related to video created one second, audio created next second, and merge failing due names mismatch which I already solved as follows:
README.md
```
#Add this lines to /var/lib/motion/motion.conf
target_dir /var/lib/motion

movie_codec mp4
movie_filename %Y%m%d/%Y%m%d_%H%M%S

on_movie_start /usr/bin/nice -n 10 ionice -c 3 /bin/bash /etc/motion/motionaudio.sh start %f
on_movie_end /bin/bash /etc/motion/motionaudio.sh stop %f

#Add this lines to /etc/sudoers
motion ALL=NOPASSWD: /usr/bin/pkill
motion ALL=NOPASSWD: /usr/bin/nohup


#install required tools and verify motion has access to run them like:
#sudo -u motion pkill
#sudo -u motion nohup
#sudo -u motion arecord
#sudo -u motion ionice
#sudo -u motion nice
#sudo -u motion ffmpeg

#if necessary add motion user to folders
#sudo chown -R motion:motion /var/lib/motion

#if nessary add execution permission to script
#sudo chmod u+x /var/lib/motion/motion-audio.sh

#check and add if neccesary groups to motion
#groups motion
# motion : motion audio video

#sudo usermod -a -G motion motion
#sudo usermod -a -G audio motion
#sudo usermod -a -G video motion

```

motion-audio.sh
```
#!/bin/bash
moviedir="/var/lib/motion/$(date +'%Y%m%d')"
action=$1
video_filename=$2
echo "Video filename ${video_filename}"
audio=$(basename "${video_filename%.*}")
aext=wav
vext=mp4
voutext=mp4

function startaudio () {
  if pgrep -f "arecord"; then sudo pkill -9 -f "arecord -f cd -r 22050 -D plughw:0,0"; fi
  echo "Iniciando grabación de audio en ${moviedir}/${audio}.${aext}"
  /usr/bin/nice -n 10 ionice -c 3 arecord -f cd -r 22050 -D plughw:0,0 "${moviedir}/${audio}.${aext}" &
  audiopid=${BASHPID}
  echo "$!" > "${moviedir}/${audio}.pid"
}

function stopaudio () {
  echo "Deteniendo grabación de audio en ${moviedir}/${audio}.${aext}"
  audiopids=$(find ${moviedir} -type f -iname "*.pid" -printf "%f\n" | cut -d. -f1)
  while read -r x; do
          echo "Matando proceso en ${x}.pid con la instruccion sudo pkill -9 -F ${x}.pid"
          if pgrep -F "${moviedir}/${x}.pid"; then sudo pkill -9 -F "${moviedir}/${x}.pid"; fi
          if [[ -f "${moviedir}/${x}.${vext}" ]] && [[ -f "${moviedir}/${x}.${aext}" ]]; then
          /usr/bin/nice -n 10 ionice -c 3 ffmpeg -y -i "${moviedir}/${x}.${vext}" -i "${moviedir}/${x}.${aext}" -c:v copy -c:a aac -b:a 128k "${moviedir}/${x}_temp.${voutext}" &&  rm -f "${moviedir}/${x}".{$aext,$vext,pid}
          mv "${moviedir}/${x}_temp.${voutext}" "${moviedir}/${x}_mixed.${voutext}"
          fi
  done <<< "${audiopids}"
  if ! pgrep -f "sudo nohup motion"; then sudo pkill -9 -f "arecord -f cd -r 22050 -D plughw:0,0"; fi
}

case ${action,,} in
stop ) stopaudio ;;
start ) startaudio ;;
esac
```

**Reply by rogerw-gh** [reply]
Hi,  I am going to try this setup - could you confirm that this will only work on a single camera per motion host - the audio stream passing through the audio device is just a single channel - running multiple sessions of motion-audio at the same time will just result in garbled, combined audio?

**Comment by jlg89** [comment]
I'll put in a note of agreement that it would be super useful to have something like an "audio_passthru" setting (takes the audio from the "netcam_url" and passes it through to the output stream), or an "audio_url" setting (takes a separate audio stream and combines it with the output audio), because motion is such a great mediator between "dumb" RTSP cameras and various home automation systems. After all, "because it's difficult" is a reasonable argument for excluding a feature, but "because something someone might do with it might be illegal someplace" really isn't. Apple, for example, doesn't seem to be concerned that HSV drops a gajillion full A/V surveillance recordings into iCloud every day.

**Comment by ceewanna** [comment]
Motion is a great tool in terms of motion detection, especially the feature in which it uses low resolution for detection and passing through high resolution to video recording. This is very efficient compared to another OpenCV app that I have also setup to compare performance. Motion is very CPU friendly while dvr-scan based on OpenCV requires substantial much more CPU or even needs GPU. The only missing piece here is voice/sound recording.

You can always argue about legal issue. In my view, if you live in the context of no-sound-recording allowed, you just get all your mp4 through ffmpeg to strip off audio. That is it. Technically, it is just an option where you leave it to user's judgment to switch it on or off. 

I live in a farm and I have Motion recording so many birds hanging around in my farm. The video recording is perfect but it's just too bad and really bad not being able to listen to their voice. One so called legal issue just kills off those lovely birds singing.

**Comment by pfzim** [comment]
Thanks for a great product!
But there is a lack of sound recording.
Please add optional pass-through audio recording from RTSP stream.

**Comment by Mr-Dave** [comment]
The code in the master branch now(Feb 05 2025) supports recording of audio from network cameras via the passthrough recording option

**Reply by pfzim** [reply]
Great news! When can we expect the release?

**Comment by BugerDread** [comment]
I have a vivotek ip7138 camera, which can stream mjpeg in better quality than mpeg4, but mpeg4 rtsp stream is with audio (i can play it with mplayer). It can stream both, so my idea is to set the mjpeg stream to highest quality to get best picture and mpeg4 to low quality just to get the audio... it will be nice when it will be possible to set it up in a way that when motion is detected in the mjpeg stream it starts to record the mpeg4 also (with audio)...

**Comment by Eloren1** [comment]
### Simplest audio settings

`nano /etc/motion/motion.conf`
```
on_event_start arecord -f S16_LE -r 22050 -D plughw:1,0 /YOURVIDEOPATH/%Y-%m-%d__%H-%M-%S.wav
on_event_end pkill -f "arecord -f S16_LE -r 22050"
```

Where `-D plughw:1,0` is an audio device to record from.

Finding your device number:  `arecord -l`, example output: `card 1: Webcam [C920 Webcam], device 0: USB Audio [USB Audio]` -> Use `plughw:1,0` (card 1, device 0).

You will get separate video and audio files. Easy!

### More info

I merge these files locally using a python script on another machine. Filenames can mismatch so I just merge them sequentially (1st mkv + 1st wav and so on)

Audio recording doesn't use the buffer, so for example I have a pre_record of 3 seconds in video, and I shift the audio file by 3 seconds when merging.

It records an extra 1 minute of audio because of the Motion's event gap logic. I keep it in a merged file as well.

**Comment by ltb427** [comment]
Based on the Motion 4.5 branch, we can implement the recording function for network cameras based on **_RTSP_**, and we can also implement the audio trigger event function.
https://github.com/ltb427/motion
